build-tools/packages/build-cli/src/commands/generate/changeset.ts

```diff
@@ -0,0 +1,263 @@
+ */
+import { Package } from "@fluidframework/build-tools";
+import { VersionBumpType } from "@fluid-tools/version-tools";
+import { Flags } from "@oclif/core";
+import chalk from "chalk";
+import humanId from "human-id";
+import { writeFile } from "node:fs/promises";
+import path from "node:path";
+import prompts from "prompts";
+
+import { BaseCommand } from "../../base";
+import { Repository, getDefaultBumpTypeForBranch } from "../../lib";
+
+const DEFAULT_BRANCH = "main";
+const INSTRUCTIONS = `
+↑/↓: Change selection
+Space: Toggle selection
+a: Toggle all
+Enter: Done`;
+
+const excludedScopes = new Set(["@fluid-example", "@fluid-internal", "@fluid-test"]);
+
+interface Choice {
+	title: string;
+	value?: Package;
+	disabled?: boolean;
+	selected?: boolean;
+	heading?: boolean;
+}
+
+export default class GenerateChangesetCommand extends BaseCommand<typeof GenerateChangesetCommand> {
+	static summary = `Generates a new changeset file.`;
+	static aliases: string[] = ["changeset:add"];
+	static enableJsonFlag = true;
+
+	static flags = {
+		branch: Flags.string({
+			char: "b",
+			description:
+				"The branch to compare against. This is used to populate the list of changed packages.",
+			default: DEFAULT_BRANCH,
+		}),
+		empty: Flags.boolean({
+			description: "Create an empty changeset file.",
+		}),
+		all: Flags.boolean({
+			description: `Include ALL packages, including examples.`,
+			default: false,
+		}),
+		uiMode: Flags.string({
+			description: `Controls the mode in which the interactive UI is displayed. The 'default' mode includes an autocomplete filter to narrow the list of packages. The 'simple' mode does not include the autocomplete filter, but has better UI that may display better in some terminal configurations. This flag is experimental and may change or be removed at any time.`,
+			default: "default",
+			options: ["default", "simple"],
+			helpGroup: "EXPERIMENTAL",
+		}),
+		...BaseCommand.flags,
+	};
+
+	static examples = [
+		{
+			description: "Create an empty changeset.",
+			command: "<%= config.bin %> <%= command.id %> --empty",
+		},
+		{
+			description: `Create a changeset interactively. Any packages edited relative to the '${DEFAULT_BRANCH}' branch will be selected by default.`,
+			command: "<%= config.bin %> <%= command.id %>",
+		},
+		{
+			description: `You can select packages relative to a different branch using --branch.`,
+			command: "<%= config.bin %> <%= command.id %> --branch next",
+		},
+		{
+			description: `By default example and private packages are excluded, but they can be included with --all.`,
+			command: "<%= config.bin %> <%= command.id %> --all",
+		},
+	];
+
+	public async run(): Promise<{
+		branch: string;
+		selectedPackages: string[];
+		changesetPath?: string;
+	}> {
+		const context = await this.getContext();
+		const { all, branch, empty, uiMode } = this.flags;
+
+		if (empty) {
+			// TODO: This only works for the root release group (client). Is that OK?
```

Comment:
If the this command/tool works only for client release group we should make it more visible.

Reply:
This only matters if we need changesets for a particular release group to go into a different folder, rather than putting everything in the root. Originally I thought we'd need them to all be separate, but I no longer think that's true. I think we can put everything in the root.

If we need to sort them, then we might need to prompt the user to select a release group in the --empty flow. But that's easy enough to add in the future if needed.

Reply:
So changesets for azure-client will still go in the root? Im curious why did you change your mind there? it seems that azure-client changesets are always specific to azure workspace.

Reply:
> So changesets for azure-client will still go in the root? Im curious why did you change your mind there? it seems that azure-client changesets are always specific to azure workspace.

There's only two release groups that matter now: client and server. Azure packages are part of the client release group now, so adding release group scope to the tool seems unnecessary right now since 99% of changes will be in client. For independent packages, we can edit the changelog directly when making changes since there's only one file to edit, and we don't batch changes. See [protocol-defs changelog](https://github.com/microsoft/FluidFramework/blob/main/common/lib/protocol-definitions/CHANGELOG.md) for an example.